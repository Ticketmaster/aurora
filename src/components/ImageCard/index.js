import React, { Children } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Card from "../Card";
import colors from "../../theme/colors";
import spacing from "../../theme/spacing";
import constants from "../../theme/constants";
import typography from "../../theme/typography";

const Image = styled.img`
  width: 100%;
  max-width: 100%;
`;

const Overlay = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-image: linear-gradient(to bottom, rgba(0, 0, 0, 0), #000000);
  display: flex;
  align-items: flex-end;
`;

const CaptionContainer = styled.div`
  position: relative;
  width: 100%;
  color: ${props => props.color};
  padding-top: 0;
  padding-bottom: 9px;
  padding-left: ${spacing.gutters.small}px;
  padding-right: ${spacing.gutters.small}px;

  @media screen and ${constants.breakpoints.mediumAndUp} {
    padding-bottom: 9px;
    padding-left: ${spacing.gutters.mediumAndUp / 2}px;
    padding-right: ${spacing.gutters.mediumAndUp / 2}px;
  }
`;

const Title = styled.span`
  display: block;
  width: 100%;
  font-size: ${typography.size.kilo};
  font-weight: ${typography.weight.semiBold};
`;

const SubTitle = styled.span`
  font-size: ${typography.size.hecto};
  font-weight: ${typography.weight.regular};
`;

const Container = styled.div`
  display: flex;
  flex-direction: column;
  position: relative;
`;

const RowContainer = Container.extend`
  flex-direction: row;
  align-items: center;
`;

const ImageCard = ({ src, alt, children, type, ...props }) => {
  const [title, subTitle, ...rest] = Children.toArray(children || []);
  const img = props.image || <Image src={src} alt={alt} />;

  if (type === "half") {
    return (
      <RowContainer>
        <Card>
          {img}
          <Overlay />
        </Card>
        <CaptionContainer color={colors.onyx.base}>
          {title}
          {subTitle}
        </CaptionContainer>
      </RowContainer>
    );
  }

  return (
    <Card>
      <Container>
        {img}
        <Overlay>
          {(title || subTitle) && (
            <CaptionContainer color={colors.white.base}>
              {title}
              {subTitle}
            </CaptionContainer>
          )}
        </Overlay>
      </Container>
      {(rest && rest.length && <div>{rest}</div>) || null}
    </Card>
  );
};

ImageCard.propTypes = {
  type: PropTypes.oneOf(["full", "half"]),
  src: PropTypes.string,
  alt: PropTypes.string,
  children: PropTypes.node,
  image: PropTypes.element
};

ImageCard.defaultProps = {
  type: "full",
  alt: "",
  src: "",
  children: null,
  image: null
};

ImageCard.Title = Title;
ImageCard.SubTitle = SubTitle;

export default ImageCard;
