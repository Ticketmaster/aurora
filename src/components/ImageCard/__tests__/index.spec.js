import React from "react";
import renderer from "react-test-renderer";

import ImageCard from "../index";

describe("ImageCard", () => {
  const title = <ImageCard.Title>Title</ImageCard.Title>;
  const subTitle = <ImageCard.SubTitle>Sub Title</ImageCard.SubTitle>;
  const children = <div>Extra content</div>;

  it("renders standard card", () => {
    const component = renderer.create(
      <ImageCard src="http://localhost/img.png" alt="image" title="image" />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders with title & subtitle", () => {
    const component = renderer.create(
      <ImageCard
        src="http://localhost/img.png"
        cardTitle={title}
        cardSubtitle={subTitle}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders with title & subtitle, extra content", () => {
    const component = renderer.create(
      <ImageCard
        src="http://localhost/img.png"
        cardTitle={title}
        cardSubtitle={subTitle}
      >
        {children}
      </ImageCard>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders with custom image", () => {
    const image = <svg />;
    const component = renderer.create(
      <ImageCard image={image} cardTitle={title} cardSubtitle={subTitle}>
        {children}
      </ImageCard>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders type = half", () => {
    const component = renderer.create(
      <ImageCard
        type="half"
        src="http://localhost/img.png"
        cardTitle={title}
        cardSubtitle={subTitle}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders with custom props", () => {
    const component = renderer.create(
      <ImageCard
        src="http://localhost/img.png"
        cardTitle={title}
        overlayProps={{ style: { display: "none" } }}
        containerProps={{ style: { display: "none" } }}
        captionContainerProps={{ style: { display: "none" } }}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders with additional content when `cardTitle` and `cardSubtitle` prop values are not provided (old API support)", () => {
    const component = renderer.create(
      <ImageCard src="http://localhost/img.png">
        {title}
        {subTitle}
        {children}
      </ImageCard>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
