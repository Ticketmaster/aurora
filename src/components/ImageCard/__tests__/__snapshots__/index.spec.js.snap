// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCard renders standard card 1`] = `
<div
  className="container-block hXHtmw"
>
  <div
    className="igiVqV"
  >
    <img
      alt="image"
      className="fIQgYS"
      src="http://localhost/img.png"
      title="image"
    />
    <div
      className="iRzHme"
    />
  </div>
</div>
`;

exports[`ImageCard renders type = half 1`] = `
<div
  className="dGrJdi"
>
  <div
    className="container-block bgkVzE"
  >
    <img
      alt=""
      className="fIQgYS"
      src="http://localhost/img.png"
      title=""
    />
  </div>
  <div
    className="brJtrF"
  >
    <span
      className="jJmwnS"
    >
      <div
        className="text text--dark text--primary image-card__title--half hTwOaP"
        disabled={false}
        size={
          Object {
            "large": "hecto",
            "medium": "hecto",
            "small": "hecto",
          }
        }
      >
        <span
          className="jJmwnS"
        >
          Title
        </span>
      </div>
    </span>
    <span
      className="eKXBpw"
    >
      <div
        className="text text--dark text--primary image-card__subtitle--half hTwOaP"
        disabled={false}
        size={
          Object {
            "large": "hecto",
            "medium": "hecto",
            "small": "hecto",
          }
        }
      >
        <span
          className="eKXBpw"
        >
          Sub Title
        </span>
      </div>
    </span>
  </div>
</div>
`;

exports[`ImageCard renders with additional content when \`cardTitle\` and \`cardSubtitle\` prop values are not provided (old API support) 1`] = `
<div
  className="container-block hXHtmw"
>
  <div
    className="igiVqV"
  >
    <img
      alt=""
      className="fIQgYS"
      src="http://localhost/img.png"
      title=""
    />
    <div
      className="iRzHme"
    >
      <div
        className="dAmSFY"
      >
        <span
          className="jJmwnS"
        >
          Title
        </span>
        <span
          className="eKXBpw"
        >
          Sub Title
        </span>
      </div>
    </div>
  </div>
  <div>
    Extra content
  </div>
</div>
`;

exports[`ImageCard renders with custom image 1`] = `
<div
  className="container-block hXHtmw"
>
  <div
    className="igiVqV"
  >
    <svg />
    <div
      className="iRzHme"
    >
      <div
        className="dAmSFY"
      >
        <span
          className="jJmwnS"
        >
          Title
        </span>
        <span
          className="eKXBpw"
        >
          Sub Title
        </span>
      </div>
    </div>
  </div>
  <div>
    Extra content
  </div>
</div>
`;

exports[`ImageCard renders with custom props 1`] = `
<div
  className="container-block hXHtmw"
>
  <div
    className="igiVqV"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <img
      alt=""
      className="fIQgYS"
      src="http://localhost/img.png"
      title=""
    />
    <div
      className="iRzHme"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <div
        className="dAmSFY"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <span
          className="jJmwnS"
        >
          Title
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`ImageCard renders with title & subtitle 1`] = `
<div
  className="container-block hXHtmw"
>
  <div
    className="igiVqV"
  >
    <img
      alt=""
      className="fIQgYS"
      src="http://localhost/img.png"
      title=""
    />
    <div
      className="iRzHme"
    >
      <div
        className="dAmSFY"
      >
        <span
          className="jJmwnS"
        >
          Title
        </span>
        <span
          className="eKXBpw"
        >
          Sub Title
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`ImageCard renders with title & subtitle, extra content 1`] = `
<div
  className="container-block hXHtmw"
>
  <div
    className="igiVqV"
  >
    <img
      alt=""
      className="fIQgYS"
      src="http://localhost/img.png"
      title=""
    />
    <div
      className="iRzHme"
    >
      <div
        className="dAmSFY"
      >
        <span
          className="jJmwnS"
        >
          Title
        </span>
        <span
          className="eKXBpw"
        >
          Sub Title
        </span>
      </div>
    </div>
  </div>
  <div>
    Extra content
  </div>
</div>
`;
